### ISSUE TITLE
incorrect padding error

### ISSUE THREAD

**Issue by yuri-0415** [wafflestudio/rookies#124]
_**binascii.Error: Incorrect padding**_ 
다음과 같은 에러가 발생하는데 해당 에러로 구글링 해봐도 해결 방법을 잘 모르겠어서 issue 남깁니다. 
postman 에서 내용을 작성하고 POST까지는 문제가 없었는데 **PUT을 시도**해보다가 해당 에러가 발생했습니다.
**이후 GET api/v1/survey를 send해도 동일한 에러가 발생**하는데(그 이전에는 발생하지 않았습니다!) 혹시 참고할 만한 방법이 있을까요?
제 추측으로는 postman에서 body와 header 등의 부분에서 잘못 작성한 것 같은데 정확한 원인을 모르겠습니다TT

에러 전문은 너무 길어서 txt 파일로 첨부했습니다. 
[error.txt](https://github.com/wafflestudio/rookies/files/5182603/error.txt)

**Comment by dodo4114** [comment]
`AttributeError: 'SessionStore' object has no attribute '_session_cache'`로 구글링 해보셨나요?

**Comment by yuri-0415** [comment]
_https://stackoverflow.com/questions/27180353/django-project-looking-for-attribute-session-cache_
1. Middleware관련 설정을 수정 2. superuser을 admin의 방법으로는 해결이 되지 않네요ㅠㅠ
postman 사용 안하고 그냥 크롬에서 실행했을 때(ex. port 8000으로 runserver)는 문제 없이 작동하는데 그러면 postman의 문제일까요?

**Comment by davin111** [comment]
이런 질문의 경우에는 반드시 본인이 실제 어떻게 request를 날리고 있는지도 스크린샷 등으로 첨부해주세요. 질문에서도 언급하셨듯이 뭔가 실수나 잘못이 포함되어 있을 수 있는데, 답변하는 입장에서는 그냥 찍어야합니다. ㅜㅜ

**Comment by davin111** [comment]
그리고 두 번째 남겨주신 메시지에서 시도하신 1. 2.를 좀 더 구체적으로 말씀해주실 수 있을까요? 특히 2.가 이해되지 않습니다.

**Comment by yuri-0415** [comment]
@davin111 
PUT을 날린 것이 아니라 단순히 GET이 된다는 것이었습니다!  이전에는 GET에서도 attribute error가 발생했는데 이유는 모르겠지만 이전의 에러는 발생하지 않고 있네요..! GET /api/v1/survey/ 를 postman에서 실행하면 되지 않지만(404 not found) 크롬에서 실행?하면 실행이 되는 상황입니다. 해당 화면은 기존 질문에 첨부했습니다.

두번째 메시지에서 시도한 2는 스택오버플로우의 내용 중 superuser를 create 했다는 뜻이었습니다. admin이라고 잘못 작성했네요..!

질문을 미흡하게 작성했는데 앞으로는 말씀해주신 것처럼 작성하겠습니다! 내용을 정확히 모르니까 질문 작성도 부족한 부분이 많았네요ㅠㅠ

**Comment by yuri-0415** [comment]
> 크롬에서 쿠키+인터넷 사용 기록을 삭제했더니 해결되었습니다!

**Comment by yuri-0415** [comment]
>> 해결된줄 알았는데 다시 같은 에러가 발생하네요ㅠㅠㅠ 다시 쿠키랑 인터넷 기록 삭제해봐도 error.txt와 동일한 에러가 발생합니다....

**Comment by davin111** [comment]
@yuri-0415 Postman에서 header, body 등에 어떤 값을 넣어보내고 있는지는 스크린샷에 담겨있지 않은 거 같은데, 포함시켜 주실 수 있을까요?

**Comment by davin111** [comment]
흠... 근데 이거 Django community에서 굉장히 최근에도 다뤄진 이슈인 거 같기도 하구요. https://code.djangoproject.com/ticket/31895 어 쩌면 Django 이슈일 수도 있을 거 같기도 한데... 아무튼 위에서 부탁드린 정보들은 제공 부탁드려요.

**Comment by yuri-0415** [comment]
@davin111 
GET 요청을 보낼때는 header, body의 내용을 변경하지 않았습니다. (스크린샷 첨부)

쿠키+인터넷 기록을 삭제했을 때에는 GET /api/v1/survey와 PUT /api/v1/user 모두 잘 작동했습니다. 그러나 이후 GET /api/v1/user를 하자마자 다시 터졌고, 그 다음부터는 쿠키+인터넷 기록을 삭제해도 해결되지 않았습니다.

환경은 윈도우 10입니다! 곧 맥을 사려고 생각 중이어서 ubuntu를 사용하지 않았는데 그래서 문제가 생긴걸까요?ㅠㅠ

다시 해보니 error message가 binascii.Error:incorrect padding error에서 binascii.Error:Invalid base64-encoded string으로 변경되어서 해당 내용을 청부합니다!
[error2.txt](https://github.com/wafflestudio/rookies/files/5188749/error2.txt)

**Comment by davin111** [comment]
yuri-0415 네 세션 관련된 거라 쿠키 같은 거 지웠을 때에 따라 영향이 있었던 것처럼 보였을 수 있었을 거 같구요, Windows 10이고, 혹시 32bit거나 다른 특이한 상황은 아닌 것인가요?

**Comment by yuri-0415** [comment]
@davin111 네 64bit이고 별다른 특이사항은 없는 것 같습니다.

**Comment by davin111** [comment]
그리고 PUT할 때 body 어떻게 보내시는지 알려주세요. 한 번에 쭉 정보를 많이 주시면 좋을 것 같습니다 ㅜㅜ 생각보다 과하게 디테일해도 좋습니다!

**Comment by davin111** [comment]
글구 위에도 했던 질문이긴 한데, 쿠키 등 인터넷 기록을 삭제하셨을 때 영향받은 것처럼 보였던 요청들은 Postman으로 날린 것이었었나요?

**Comment by yuri-0415** [comment]
@davin111 

body에서 raw data로 이런 식으로 주었었습니다.
{
    "username" : "yuri",
    "email" : "yuri@naver.com",
    "password" : "0415"
}

네 postman으로 날렸었습니다! 모두 postman으로 날렸을 때 GET, PUT 모두 작동하다가 GET /api/v1/user을 보내고 이후부터는 작동하지 않았습니다.

**Comment by davin111** [comment]
이건 그냥 무관한 거일 수도 있는데, form-data 형식이 아니라 raw로 body를 보내고 계신데요. 이는 무관하긴 한데, 이렇게 한다면 text가 아닌 json을 택해야할 것 같습니다.

**Comment by yuri-0415** [comment]
아! PUT 보낼때는 headers에 content-type  / application-json을 포함해주었습니다! 다른 hidden headers는 바꾸지 않았었구요

**Comment by davin111** [comment]
네 이거는 지금 겪으시는 문제 상황과 무관하긴 한 거 같은데, 그러니까 raw에 해당하는 부분도 저렇게 JSON으로 택해져서 header의 Content-Type이 그렇게 자동으로 설정되었다는 말씀이신 것이지요?

**Comment by yuri-0415** [comment]
![image](https://user-images.githubusercontent.com/63419338/92488149-22d22600-f229-11ea-9da4-91ec05711aaa.png)

이런식으로 자동 생성된 hidden headers가 있었고 Content-type / application-json을 설정해주었었습니다!

**Comment by davin111** [comment]
PUT 보낼 때의 hidden headers도 이 상태에서 그대로 펼쳐서 캡처주시고, 저 Content-Type은 raw에서 JSON을 택하시면 이렇게 수동으로 택하실 필요가 없습니다.

**Comment by yuri-0415** [comment]
아하! 그렇군요
![image](https://user-images.githubusercontent.com/63419338/92488452-852b2680-f229-11ea-8fd9-3104ad607d95.png)

**Comment by davin111** [comment]
저 Content-Type 수동으로 설정하느라고 따로 이상하게 빠져있는 거 같은데, 지워주시고 raw의 JSON 통해서 자동 설정 잘 되게 만들어주시겠어요? 이게 딱히 원인일 거라고 생각하진 않지만, 잘 모르겠는 상황에서 부차적인 변수들 다 제외하면 좋겠네요.

**Comment by davin111** [comment]
음 글구 또 하나 궁금한 것이, 정상적으로 PUT 요청 날리려면 header에 X-CSRFToken 설정되어있어야할 거 같은데, 이거 없어도 잘 되시던가요?

**Comment by davin111** [comment]
header 정리하실 때, 지금 header가 11로 되어있는데, 실제로 체크걸려있는 건 9개에요, 공백 같은 거 이상하게 포함된 거 없는지 잘 체크하시고 깔끔하게 정리해주세요. 이건 유관할 수도 있습니다. 그래서 꼬치꼬치 여쭈면서 계속 스크린샷 부탁드린 것이구요. 앞으로는 꼭 부탁드립니다!

**Comment by davin111** [comment]
혹시 그리고 127.0.0.1로 하든, localhost로 하든 마찬가지 현상인지도 한 번 테스트해보시면 좋을 거 같아요!

**Comment by yuri-0415** [comment]
![image](https://user-images.githubusercontent.com/63419338/92488986-33cf6700-f22a-11ea-9bb3-d1506d3b629d.png)

말씀해주신 내용 모두 반영했습니다. PUT 요청을 보내봤을때 header가 12개로 뜹니다.  여기서 header는 모두 건드리지 않았고 자동생성되었습니다. 쿠키에 포함된 csrftoken이 아니라 별도의 X-CSRFToken이 생겨야 한다는 말씀이신거죠? 그 부분은 생성되지 않는 듯합니다. 에러 메시지는 변함이 없습니다!

넵 알겠습니다! 번거롭게 해서 죄송하네요ㅜㅜ

**Comment by davin111** [comment]
혹시 그리고 127.0.0.1로 하든, localhost로 하든 마찬가지 현상인지도 한 번 테스트해보시면 좋을 거 같아요!

**Comment by davin111** [comment]
X-CSRFToken은 직접 넣으셔야하고(#105 등등 참고), 그렇게 하지 않았을시 회원가입이나 로그인하신 후라면, 403 CSRF 관련 에러가 나타나야했을텐데, 그렇지 않다는 점이 일단 의문스럽네요.

**Comment by yuri-0415** [comment]
localhost일 경우 error가 binascii.Error: Incorrect padding
127.0.0.1일 경우에 binascii.Error:Invalid base64-encoded string이 뜨는듯 합니다!

**Comment by davin111** [comment]
혹시 Django MIDDLEWARE 설정 직접 건들리신 거 있으신가요?

**Comment by davin111** [comment]
그리고 https://github.com/wafflestudio/rookies/issues/124#issuecomment-688914071 여기서 주신 스크린샷 봤을 때, Content-Type이 text인걸 봐서는, raw를 JSON으로 고치시지 않은 것 같은데요 ㅜㅜ

**Comment by davin111** [comment]
이전에 PUT으로 문제가 없이 동작했다, 라고 하셨을 때는 회원가입 API(이게 본문에서 언급하신, 잘 동작했다는 POST API인가요?) 이후 X-CSRFToken을 명시하지 않았는데도 403이 아니라 정상 동작을 했다는 것인가요? 그리고 이전에도 이후에도 계속 X-CSRFToken은 넣으신 적 없었구요.

**Comment by davin111** [comment]
??? 방금 https://github.com/wafflestudio/rookies/issues/124#issuecomment-688915305 이 코멘트 수정하면서 추가된 내용인 `settings.py의 SECRET_KEY에는 csrf token가 들어있습니다. 그 외로 별도 설정할 게 있는 건가요?`을 보았는데, settings.py의 SECRET_KEY를 직접 수정하신 것인가요? 일단 그러면 안 됩니다...

**Comment by yuri-0415** [comment]
세미나에서 secret key를 수정하신거로 기억했는데 아닌건가요?ㅠㅠ

**Comment by davin111** [comment]
네 아니구요. 위에도 언급드렸듯이, #105 등 CSRF 방어 관련해서 해야할 일은 여러 곳에서 여러 번 설명드렸습니다. [과제 내용](https://github.com/wafflestudio/rookies/blob/master/backend/seminar1/assignment.md#과제-내용)에서도 저 이슈 링크 드렸구요. 그리고 세미나 1의 영상 등도 올라와있는데, 많이 너무하시네요... ㅜㅜ SECRET_KEY를 일단 수정할 거면 이게 뭔지 [공식 문서](https://docs.djangoproject.com/en/2.2/ref/settings/#secret-key) 등을 찾아볼 필요도 있었을 것 같구요.

**Comment by yuri-0415** [comment]
middleware설정은 동일합니다(구글링했을때 특정 middleware를 제외하라는 이야기가 있어서 제외했었습니다. 그런데 반드시 포함시켜야 한다는 error메시지가 나와서 바로 기존과 동일하게 복구했습니다)

그 다음 코멘트는 반영하겠습니다!

정말 죄송합니다ㅠㅠㅠㅠㅠㅠ

**Comment by davin111** [comment]
일단 여기서 대화 나눈 내용과 다른 이슈들, 그리고 과제 내용 한 번씩 다시 살펴보시면서 나머지 부분을 올바르게 하신 후, 그래도 문제가 재현되는지 아닌지 남겨주시면 될 거 같습니다. 일단 SECRET_KEY 직접 수정하신 것은 치명적으로 보입니다. 또한 제가 위에 질문드렸던 여러가지도 답은 부탁드리구요.

**Comment by davin111** [comment]
글구 SECRET_KEY 관련 [공식 문서](https://docs.djangoproject.com/en/2.2/ref/settings/#secret-key)에도 나와있듯, 이는 session 등 여러가지 내용을 hashing(암호화 정도로 생각하셔도 될듯)할 때 사용하게 되는데요, 아무래도 이것을 임의로 수정하신 것이 원인이었던 것으로 보입니다. 결국 원인을 찾기는 찾은 것 같네요. 그런데 제가 엄청나게 digging 해서야 가능했었고, 앞으로는 이런 방식은 아니면 좋겠습니다... 시간 되게 많이 써야했어요 ㅜ

**Comment by yuri-0415** [comment]
네 정말 죄송합니다ㅠㅠ 제실수로 잘못 기억한 부분인데 세미나 장님께서 너무 많은 시간을 쏟으셔서 죄송할 따름입니다ㅠㅠㅠ 우선 말씀해주신 부분 반영해서 수정해보겠습니다... 앞으로는 이런 일 없도록 노력하겠습니다...ㅠㅠ 함께 고민해주셔서 감사합니다!

**Comment by yuri-0415** [comment]
migrate할 때 잘못된 setting(secret keyㅠㅠㅠ😭)으로 보내서 이후 논의 사항을 반영해도 binascii error가 해결되지는 않더라구요! 아마 제가 중간에 이것저것 건드린 부분이 문제가 되어 충돌이 나는 것 같습니다. 그래서 처음으로 돌아가 다시 migrate하는 방식으로 해결했습니다. 반성하고 앞으로는 주의하겠습니다...! 다시 한번 도와주셔서 감사해요😉

**Comment by davin111** [comment]
@yuri-0415 오, 어쨌든 해당 문제는 완전히 해결되셨나요! 다행이네요.
